File: tvcast/cast/launcher.py
"""Tell a TV to play a URL. The MVP ships the DLNA AVTransport launcher.

A launcher has three methods: play(url, title), stop(), state(). Future transports
(Cast, Roku ECP, ADB intents) implement the same three.

XML note: responses are parsed with the stdlib ElementTree. They come from a device on
the LAN that we chose to talk to, and expat does not resolve external entities, so the
usual XXE concern does not apply; a hostile renderer could at worst waste our CPU.
"""
import re
import urllib.error
import urllib.request
import xml.etree.ElementTree as ET
from xml.sax.saxutils import escape

SERVICE_NS = "urn:schemas-upnp-org:service:{service}:1"
STATES = {"PLAYING", "STOPPED", "TRANSITIONING", "PAUSED_PLAYBACK", "NO_MEDIA_PRESENT"}


class UpnpError(Exception):
    def __init__(self, code, description, http_status=None):
        super().__init__(f"UPnP error {code}: {description}" if code else description)
        self.code, self.description, self.http_status = code, description, http_status


def soap_envelope(service, action, args):
    ns = SERVICE_NS.format(service=service)
    body = "".join(f"<{k}>{escape(str(v))}</{k}>" for k, v in args.items())
    return ('<?xml version="1.0" encoding="utf-8"?>'
            '<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" '
            's:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body>'
            f'<u:{action} xmlns:u="{ns}">{body}</u:{action}></s:Body></s:Envelope>')


def _local(tag):
    return tag.split("}", 1)[-1]


def parse_soap_response(body):
    """Return the child elements of the first *Response element as {name: text}."""
    out = {}
    try:
        root = ET.fromstring(body)
    except ET.ParseError:
        return out
    for elem in root.iter():
        if _local(elem.tag).endswith("Response"):
            for child in elem:
                out[_local(child.tag)] = (child.text or "").strip()
            break
    return out


def parse_upnp_fault(body):
    code = re.search(r"<errorCode>(\d+)</errorCode>", body)
    desc = re.search(r"<errorDescription>(.*?)</errorDescription>", body, re.S)
    return (int(code.group(1)) if code else None,
            desc.group(1).strip() if desc else body[:200])


def didl_lite(url, title, mime="video/mpeg"):
    """Metadata for SetAVTransportURI. Some renderers refuse to play without it."""
    return ('<DIDL-Lite xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" '
            'xmlns:dc="http://purl.org/dc/elements/1.1/" '
            'xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" '
            'xmlns:dlna="urn:schemas-dlna-org:metadata-1-0/">'
            f'<item id="0" parentID="-1" restricted="1"><dc:title>{escape(title)}</dc:title>'
            '<upnp:class>object.item.videoItem</upnp:class>'
            f'<res protocolInfo="http-get:*:{mime}:DLNA.ORG_OP=00;DLNA.ORG_CI=0;'
            'DLNA.ORG_FLAGS=01700000000000000000000000000000">'
            f'{escape(url)}</res></item></DIDL-Lite>')


class SoapClient:
    def __init__(self, control_url, opener=urllib.request.urlopen, timeout=10):
        self.control_url, self.opener, self.timeout = control_url, opener, timeout

    def call(self, service, action, args):
        ns = SERVICE_NS.format(service=service)
        req = urllib.request.Request(
            self.control_url, data=soap_envelope(service, action, args).encode("utf-8"),
            method="POST",
            headers={"Content-Type": 'text/xml; charset="utf-8"',
                     "SOAPAction": f'"{ns}#{action}"'})
        try:
            with self.opener(req, timeout=self.timeout) as resp:
                return parse_soap_response(resp.read().decode("utf-8", "replace"))
        except urllib.error.HTTPError as e:
            body = e.read().decode("utf-8", "replace")
            code, desc = parse_upnp_fault(body)
            raise UpnpError(code, desc, http_status=e.code) from None
        except (urllib.error.URLError, OSError) as e:
            raise UpnpError(None, f"cannot reach {self.control_url}: {e}") from None


class DlnaLauncher:
    name = "dlna"

    def __init__(self, control_url, client=None):
        self.control_url = control_url
        self.client = client or SoapClient(control_url)

    def _avt(self, action, **args):
        return self.client.call("AVTransport", action, {"InstanceID": 0, **args})

    def play(self, url, title="Mac screen"):
        try:
            self._avt("Stop")
        except UpnpError as e:
            if e.code is None:
                raise  # unreachable, no point continuing
            # Some renderers fault on Stop when idle; that is fine.
        self._avt("SetAVTransportURI", CurrentURI=url, CurrentURIMetaData=didl_lite(url, title))
        self._avt("Play", Speed="1")

    def stop(self):
        self._avt("Stop")

    def nudge(self):
        """Re-issue Play on an already-playing renderer: a control command with no visible
        effect that some TVs count as activity, holding off their screensaver."""
        self._avt("Play", Speed="1")

    def state(self):
        try:
            s = self._avt("GetTransportInfo").get("CurrentTransportState", "UNKNOWN")
        except UpnpError:
            return "UNKNOWN"
        return s if s in STATES else "UNKNOWN"

    def position(self):
        try:
            return self._avt("GetPositionInfo").get("RelTime", "?")
        except UpnpError:
            return "?"

File: macapp/Sources/TVCastKit/Launcher.swift
import Foundation

public enum TransportState: String, Sendable {
    case playing = "PLAYING"
    case stopped = "STOPPED"
    case transitioning = "TRANSITIONING"
    case pausedPlayback = "PAUSED_PLAYBACK"
    case noMedia = "NO_MEDIA_PRESENT"
    case unknown = "UNKNOWN"
}

public struct UPnPError: Error, CustomStringConvertible {
    public let code: Int?
    public let detail: String
    public var description: String {
        code.map { "UPnP error \($0): \(detail)" } ?? detail
    }
}

/// Speaks DLNA AVTransport to a TV: play a URL, stop, read state, nudge. Mirrors the
/// Python DlnaLauncher so behaviour against real TVs stays identical.
public struct DlnaLauncher {
    public let controlURL: URL
    let timeout: TimeInterval

    public init(controlURL: URL, timeout: TimeInterval = 10) {
        self.controlURL = controlURL
        self.timeout = timeout
    }

    static let service = "urn:schemas-upnp-org:service:AVTransport:1"

    // MARK: control

    public func play(url: String, title: String = "Mac screen") throws {
        // Some renderers fault on Stop when idle; that is fine, but a network failure is not.
        do { try call("Stop", ["InstanceID": "0"]) }
        catch let e as UPnPError where e.code == nil { throw e }
        catch {}
        try call("SetAVTransportURI", [
            "InstanceID": "0",
            "CurrentURI": url,
            "CurrentURIMetaData": Self.didlLite(url: url, title: title),
        ])
        try call("Play", ["InstanceID": "0", "Speed": "1"])
    }

    public func stop() throws {
        try call("Stop", ["InstanceID": "0"])
    }

    /// Re-issue Play on an already-playing renderer: a control command with no visible
    /// effect that some TVs count as activity, holding off their screensaver.
    public func nudge() throws {
        try call("Play", ["InstanceID": "0", "Speed": "1"])
    }

    public func state() -> TransportState {
        guard let resp = try? call("GetTransportInfo", ["InstanceID": "0"]),
              let raw = resp["CurrentTransportState"] else { return .unknown }
        return TransportState(rawValue: raw) ?? .unknown
    }

    public func position() -> String {
        (try? call("GetPositionInfo", ["InstanceID": "0"]))?["RelTime"] ?? "?"
    }

    // MARK: SOAP

    @discardableResult
    func call(_ action: String, _ args: [(String, String)]) throws -> [String: String] {
        let body = args.map { "<\($0.0)>\(Self.escape($0.1))</\($0.0)>" }.joined()
        let envelope = """
        <?xml version="1.0" encoding="utf-8"?>\
        <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" \
        s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body>\
        <u:\(action) xmlns:u="\(Self.service)">\(body)</u:\(action)></s:Body></s:Envelope>
        """
        var req = URLRequest(url: controlURL, timeoutInterval: timeout)
        req.httpMethod = "POST"
        req.setValue("text/xml; charset=\"utf-8\"", forHTTPHeaderField: "Content-Type")
        req.setValue("\"\(Self.service)#\(action)\"", forHTTPHeaderField: "SOAPAction")
        req.httpBody = envelope.data(using: .utf8)

        let sem = DispatchSemaphore(value: 0)
        var data: Data?
        var status = 0
        var transportError: Error?
        URLSession.shared.dataTask(with: req) { d, resp, err in
            data = d
            status = (resp as? HTTPURLResponse)?.statusCode ?? 0
            transportError = err
            sem.signal()
        }.resume()
        _ = sem.wait(timeout: .now() + timeout + 1)

        if let transportError {
            throw UPnPError(code: nil, detail: "cannot reach \(controlURL): \(transportError.localizedDescription)")
        }
        let text = data.map { String(decoding: $0, as: UTF8.self) } ?? ""
        if status >= 400 || status == 0 {
            let fault = Self.parseFault(text)
            throw UPnPError(code: fault.0, detail: fault.1)
        }
        return Self.parseResponse(text)
    }

    // MARK: helpers (overload with array so call sites can pass ordered args)

    func call(_ action: String, _ args: [String: String]) throws -> [String: String] {
        // Ordered so envelopes are stable and testable (InstanceID first).
        let order = ["InstanceID", "CurrentURI", "CurrentURIMetaData", "Speed", "Unit", "Target"]
        let sorted = args.sorted { a, b in
            (order.firstIndex(of: a.key) ?? 99) < (order.firstIndex(of: b.key) ?? 99)
        }
        return try call(action, sorted.map { ($0.key, $0.value) })
    }

    static func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "&", with: "&amp;")
         .replacingOccurrences(of: "<", with: "&lt;")
         .replacingOccurrences(of: ">", with: "&gt;")
         .replacingOccurrences(of: "\"", with: "&quot;")
    }

    static func didlLite(url: String, title: String, mime: String = "video/mpeg") -> String {
        """
        <DIDL-Lite xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" \
        xmlns:dc="http://purl.org/dc/elements/1.1/" \
        xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" \
        xmlns:dlna="urn:schemas-dlna-org:metadata-1-0/">\
        <item id="0" parentID="-1" restricted="1"><dc:title>\(escape(title))</dc:title>\
        <upnp:class>object.item.videoItem</upnp:class>\
        <res protocolInfo="http-get:*:\(mime):DLNA.ORG_OP=00;DLNA.ORG_CI=0;\
        DLNA.ORG_FLAGS=01700000000000000000000000000000">\
        \(escape(url))</res></item></DIDL-Lite>
        """
    }

    static func parseResponse(_ body: String) -> [String: String] {
        let p = ResponseParser()
        return p.parse(body)
    }

    static func parseFault(_ body: String) -> (Int?, String) {
        func between(_ open: String, _ close: String) -> String? {
            guard let a = body.range(of: open), let b = body.range(of: close, range: a.upperBound..<body.endIndex)
            else { return nil }
            return String(body[a.upperBound..<b.lowerBound])
        }
        let code = between("<errorCode>", "</errorCode>").flatMap { Int($0.trimmingCharacters(in: .whitespaces)) }
        let desc = between("<errorDescription>", "</errorDescription>")?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (code, desc ?? String(body.prefix(200)))
    }
}

/// Pulls the child elements of the first *Response element into a dictionary.
final class ResponseParser: NSObject, XMLParserDelegate {
    private var out: [String: String] = [:]
    private var inResponse = false
    private var key: String?
    private var text = ""

    func parse(_ body: String) -> [String: String] {
        guard let data = body.data(using: .utf8) else { return [:] }
        let parser = XMLParser(data: data)
        parser.delegate = self
        parser.parse()
        return out
    }

    private func local(_ name: String) -> String {
        name.contains(":") ? String(name.split(separator: ":").last!) : name
    }

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?,
                qualifiedName qName: String?, attributes attributeDict: [String: String]) {
        let tag = local(elementName)
        if tag.hasSuffix("Response") { inResponse = true; return }
        if inResponse { key = tag; text = "" }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) { text += string }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?,
                qualifiedName qName: String?) {
        let tag = local(elementName)
        if tag.hasSuffix("Response") { inResponse = false; key = nil; return }
        if inResponse, let k = key, k == tag {
            out[k] = text.trimmingCharacters(in: .whitespacesAndNewlines)
            key = nil
        }
    }
}
